tests/component/_digital_utils.py

```diff
@@ -133,19 +150,56 @@ def _get_waveform_data_msb(waveform: DigitalWaveform) -> list[int]:
     return [_bool_array_to_int_msb(sample) for sample in waveform.data]
 
 
-def _create_digital_waveform(num_samples: int, num_lines: int = 1) -> DigitalWaveform:
-    waveform = DigitalWaveform(num_samples, num_lines)
+def _create_digital_waveform(
+    num_samples: int,
+    num_lines: int = 1,
+    invert: bool = False,
+    dtype: type[numpy.generic] = numpy.uint8,
+) -> DigitalWaveform:
+    waveform = DigitalWaveform(num_samples, num_lines, dtype=dtype)
     expected_data = _get_digital_data(num_lines, num_samples)
+    _set_waveform_data(num_samples, num_lines, waveform, expected_data, invert=invert)
+    return waveform
+
+
+def _create_waveform_for_line(num_samples: int, line_number: int) -> DigitalWaveform:
+    waveform = DigitalWaveform(num_samples, 1)
+    expected_data = _get_expected_data_for_line(num_samples, line_number)
+    _set_waveform_data(num_samples, 1, waveform, expected_data)
+    return waveform
 
+
+def _create_waveform_for_lines(
+    num_samples: int, first_line: int, num_lines: int, dtype: type[numpy.generic]
+) -> DigitalWaveform:
```

Comment:
`DigitalWaveform` has an implicit `Any`, use `DigitalWaveform[Any]` or `DigitalWaveform[np.bool]` instead.

This only applies to type hints, not constructor calls.